### ISSUE TITLE
WrappedArray[Byte].hashCode() method works differently from List[Char].hashCode() and violates equals/hashCode contract

### ISSUE THREAD

**Issue by yeputons** [scala/bug#10690]
## Code

```scala
val x = Seq[Byte](10)  // List(10)
val y = Array[Byte](10).toSeq  // WrappedArray.ofByte
x.hashCode() // 583502619
y.hashCode() // -61381922
x == y // true
x.equals(y) // true
x.hashCode() == y.hashCode() // false, contract violated!
```

**Expected behavior**: if `x == y`, then `x.hashCode() = y.hashCode()`

**Real behavior**: `x == y`, but `x.hashCode() != y.hashCode()`

## Details

[Snippet on Scastie](https://scastie.scala-lang.org/3yhfklwERrifhcbNKwcvsQ). Note that `hashCode` works in the same way for all other versions of `Seq`s and `Array`s listed there (e.g. `Char`, `Int`, `Float`...), it's `Array[Byte]` which calculates `hashCode` differently.

I've also tested it locally with Scala version 2.12.4 and Oracle Java 1.8.0_151 x86 on Windows 7 En.

## Real world implication

`Map[Seq[Byte], T]` (which uses hash map internally) may break if one uses both `Array[Byte]` and something else as keys. In particular, it may be impossible to retrieve `Seq[Byte](10)` if one uses `Array[Byte](10)` as a key—that was how I found the issue in the first place.

E.g. the following snippet evaluates to `Some(1)` on my system if run with `n = 4`, but `None` if run with `n = 5`. Your system may have different behavior, so try tweaking `n`.

```scala
val n = 4
val m = (1 to n).map { x => Seq(x.toByte) -> x }.toMap
m.get(Array[Byte](1).toSeq)
```

## Probable cause

`WrappedArray[Byte]` is implemented via [`WrappedArray.ofByte`][1] which calls `MurmurHash3.wrappedBytesHash(array)` to calculate `hashCode` (which in turn calls [`bytesHash`][3]). Other `WrappedByte` instances (e.g. [`WrappedArray.ofChar`][2]) call `MurmurHash3.wrappedArrayHash(array)` instead (which calls [`arrayHash`][4]). The code is different and the result apparently is different as well.

## Suggested fixes

1. Call `wrappedArrayHash` in `WrappedArray.ofByte` as well.
2. Document that `MurmurHash3.bytesHash` may produce a result different from that of `MurmurHash3.arrayHash` even if called from the same array.

[1]: https://github.com/scala/scala/blob/68d66968eede247adf6d784e0d8f0f343d9124d4/src/library/scala/collection/mutable/WrappedArray.scala#L135
[2]: https://github.com/scala/scala/blob/68d66968eede247adf6d784e0d8f0f343d9124d4/src/library/scala/collection/mutable/WrappedArray.scala#L159
[3]: https://github.com/scala/scala/blob/68d66968eede247adf6d784e0d8f0f343d9124d4/src/library/scala/util/hashing/MurmurHash3.scala#L131
[4]: https://github.com/scala/scala/blob/68d66968eede247adf6d784e0d8f0f343d9124d4/src/library/scala/util/hashing/MurmurHash3.scala#L118

## Potential problem

1. If someone stores hash codes of `Array[Byte]` on a disk, they suddenly no longer make sense.
2. If an app uses two versions of Scala (with and without a fix), it's possible that `Map[Seq[Byte], T]` from one part of the program should not be used in the other without breaking some invariants. Moreover, it's not validated in compile-time.

**Comment by ataraxer** [comment]
(Disclaimer: just passing through, not a Scala contributor.)

I woulds strongly suggest against using mutable data structures as map keys, because modification of such structures may lead to inability to lookup corresponding values if it's `hashCode` changes. For historic reasons `Seq` can be implemented as both mutable and immutable collections, even though it only exposes immutable methods. Hence it is perfectly valid for `WrappedArray[T]` to implement `toSeq` as `def toSeq: Seq[T] = this` which is (basically) what happens under the hood.

I would suggest using `scala.collection.immutable.Seq`, which can be constructed from other collections by using method `to`:
```scala
import scala.collection.immutable
myCollection.to[immutable.Seq]
```
Though you should be aware that it will lead to additional collection copy. If your use-case is limited to sequences of bytes you can use something akin to [`akka.util.ByteString`](https://doc.akka.io/api/akka/current/akka/util/ByteString.html). It still copies data by default, to prevent possible mutations, but if you are the sole owner of particular byte array and can guarantee that it is not going to be modified you can wrap it with [`akka.util.ByteString.fromArrayUnsafe`](https://doc.akka.io/api/akka/current/akka/util/ByteString$.html#fromArrayUnsafe(array:Array[Byte]):akka.util.ByteString), which wraps array directly.

This inconsistent behaviour is most likely caused by performance and/or backwards compatibility concerns. Note that it is fixed in upcoming standard library rewrite for new [`ImmutableArray`](https://github.com/scala/collection-strawman/blob/master/collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala#L195) wrapper, but not for [`WrappedArray`](https://github.com/scala/collection-strawman/blob/master/collections/src/main/scala/strawman/collection/mutable/WrappedArray.scala#L136).

**Comment by joshlemer** [comment]
@SethTisue @julienrf Would changing this behavior to a consistent hash in `scala.collection.immutable.ArraySeq` be too much of a breaking change for 2.13?

**Comment by julienrf** [comment]
The code given in the issue description behaves correctly in 2.13.x (the hashcodes are the same).

**Comment by joshlemer** [comment]
@julienrf are you sure, this test fails on 2.13.x for me
```scala
  @Test
  def t10690 : Unit = {
    val x = Seq[Byte](10)  
    val y = Array[Byte](10).toSeq  
    assertEquals(x.hashCode(), y.hashCode()) 
  }
```

**Comment by julienrf** [comment]
Indeed, I was 2 days behind 2.13.x. Something [here](https://github.com/scala/scala/compare/0fd1b66...2.13.x#diff-a7ab156ecfe37114bc259ba947342e46) has broken the hashcode.

**Comment by NthPortal** [comment]
It's probably a0f7c1b4dfcc9b401ed1ddc7727c8bb7fe1c18dc, which changes `immutable.ArraySeq.ofByte.hashCode` to use `MurmurHash3.byteArraySeqHash`. It should be noted that `mutable.ArraySeq.ofByte.hashCode` already used that.

**Comment by smishmash** [comment]
Are there any plans to port this fix to 2.12? We use byte sequences as keys, and this is giving us serious problems.

**Comment by SethTisue** [comment]
There is no such plan. A community-submitted pull request with a backport would be accepted if it seemed sufficiently compatible and safe. (I'm wording that cautiously because the 2.12 collections are much different than the 2.13 collections and I haven't assessed what the relevant code looks like in 2.12. Doing that assessment would be part of the work of doing the backport.)
